src/witan/send/adroddiad/ncy.clj

```diff
@@ -1,7 +1,298 @@
+   year, using the first such pair (with smallest census/calendar year)
+   as a reference point from which to calculate the nil NCYs, assuming
+   the CYP progresses at a rate of 1 NCY per census/calendar-year.
+
+   For rows where imputation is required (nil? NCY) and possible the
+   imputed NCY value is returned in the specifiec column and a string
+   describing the imputation mapped into a separate column.
+
+   For rows where imputation is not requried (some? NCY) or not possible
+   the existing NCY value (which may be nil) is returned in the specified
+   column and the imputation description column is left nil.
+
+   Names of columns of `ds` to use are specified in (optional) options
+   map (or as keyword options) as follows:
+   `id`             - unique identifier for the pupil
+                      {default :id}
+   `cy`             - calendar|census year
+                      {default :calendar-year}
+   `ncy`            - National Curriculum Year
+                      {default :academic-year}
+   `ncy-imputed`    - column to contain NCYs after imputation
+                      (can be the same as `ncy`)
+                      {default :academic-year}
+   `ncy-imputation` - imputation description column in returned dataset
+                      {default :academic-year-imputation}
+
+  NOTE: Row order of ds may NOT be preserved.
+  NOTE: Algorithm relies on input dataset `ds` NOT being named \"_ref\".
+  "
+  ([ds & {:keys [id cy ncy ncy-imputed ncy-imputation]
+          :or   {id             :id
+                 cy             :calendar-year
+                 ncy            :academic-year
+                 ncy-imputed    :academic-year
+                 ncy-imputation :academic-year-imputation}}]
+   (if (zero? (-> ds
+                  (tc/drop-missing [id cy ncy])
+                  (tc/row-count)))
+     (-> ds
+         (tc/add-column ncy-imputation nil))
+     (-> ds
+         (tc/left-join (-> ds
+                           (tc/select-columns [id cy ncy])
+                           (tc/drop-missing)
+                           (tc/order-by       [id cy ncy])
+                           (tc/group-by       [id       ])
+                           (tc/first)
+                           (tc/ungroup)
+                           (tc/rename-columns {cy  :cy-ref
+                                               ncy :ncy-ref})
+                           (tc/set-dataset-name "_ref"))
+                       [id])
+         (tc/map-columns ncy-imputation [ncy cy :ncy-ref :cy-ref]
+                         (fn [ncy cy ncy-ref cy-ref]
+                           (when (and (nil? ncy) (some? cy) (some? ncy-ref) (some? cy-ref))
+                             (str "Imputed from NCY=" ncy-ref " in " cy-ref ))))
+         (tc/map-columns ncy-imputed    [ncy cy :ncy-ref :cy-ref ncy-imputation]
+                         (fn [ncy cy ncy-ref cy-ref ncy-imputation]
+                           (if (and (nil? ncy) (some? ncy-imputation))
```

Comment:
again, probably don't need `some?`

Reply:
Left this one as `(some? ncy-imputation)` as I wanted it to read explicitly as "some NCY imputation possible?"